appengine/pond_storage/rank.py

```diff
@@ -15,72 +15,165 @@
+        'entry_key': entry.key.urlsafe(),
+        'js': entry.duck_key.get().code.js
+    })
+    if len(duck_list) == 4:
+      break
+  # 3. Verify whether enough entries could be found for a match
+  if len(duck_list) != 4:
+    logging.info('Not enough opponents available for match.')
+    return None
+  # 4. Store match request in datastore.
+  match = MatchRequest(entry_keys=entry_keys)
+  match_key = match.put()
+  match_key_urlsafe = match_key.urlsafe()
+  logging.info('Created match request. matchKey:%s', match_key_urlsafe)
+  # 5. Return match request information.
+  return {'duck_list': duck_list, 'match_key': match_key_urlsafe}
+
+def cleanup_expired_matches():
+  """Deletes expired pending match requests in datastore."""
+  now = datetime.datetime.now()
+  expired_requests_keys = MatchRequest.query(
+      MatchRequest.expiry_time < now).fetch(keys_only=True)
+  for request_key in expired_requests_keys:
+    request_key.delete()
+    logging.info('Cleaned up expired match with key:%s', request_key.urlsafe())
+
+def validate_match_result(match_key, entry_keys_urlsafe):
+  """Returns whether the given match result combination is valid."""
+  match_request = match_key.get()
+  if match_key.kind() != 'MatchRequest' and match_request:
+    return False
+  request_keys_urlsafe = {key.urlsafe() for key in match_request.entry_keys}
+  return request_keys_urlsafe == set(entry_keys_urlsafe)
+
+@ndb.transactional(xg=True)
+def apply_match_result(entry_keys_urlsafe):
+  """Applies match result if possible and returns True on success."""
+  leaderboard_entries = []
+  rankings = []
+  # Extract the ranking information.
+  for entry_key_urlsafe in entry_keys_urlsafe:
+    entry_key = ndb.Key(urlsafe=entry_key_urlsafe)
+    entry = entry_key.get()
+    if entry:
+      leaderboard_entries.append(entry)
+      rankings.append(entry.ranking)
+    else:
+      logging.error('Entry not found. entryKey:%s', entry_key)
+      return False
+  # Apply new duck rankings.
+  rankings.sort()
+  for i, entry in enumerate(leaderboard_entries):
+    entry.update_ranking(rankings[i])
+  return True
+
+def process_match_result(match_key, entry_keys_urlsafe):
+  """Validates and applies the given match results, returns True on success."""
+  if validate_match_result(match_key, entry_keys_urlsafe):
+    match_key.delete()
+    logging.info('Deleted match request. matchKey:%s', match_key_urlsafe)
+    if apply_match_result(entry_keys_urlsafe):
+      logging.info('Applied match results. matchKey:%s', match_key_urlsafe)
+      return True
+    else:
+      logging.error(
+          'Processed match results, not applied. matchKey:%s entryKeys:%s',
+          match_key_urlsafe, entry_keys_urlsafe)
+  else:
+    logging.error(
+        'No match request found for given result. matchKey:%s entryKeys:%s',
+        match_key_urlsafe, entry_keys_urlsafe)
+  return False
+
+forms = cgi.FieldStorage()
+if forms.has_key('matchKey') and forms.has_key('entryKeys'):
```

Comment:
Still trying to figure out how, but I think we should make our decision here based on whether the http method is a POST or a GET. I think it will make it more clear what is happening and also easier to debug.